app/src/main/java/edu/wisc/ece/pinpoint/utils/FirebaseDriver.java

```diff
@@ -120,6 +121,28 @@ public Task<Pin> fetchPin(@NonNull String pid) {
         });
     }
 
+    public Pin getCachedPin(@NonNull String pid) {
+        return pins.get(pid);
+    }
+
+    public Task<String> dropPin(@NonNull String content, @NonNull Pin.PinType type,
+                                @NonNull Location location, String caption) {
+        Map<String, Object> data = new HashMap<>();
+        data.put("content", content);
+        data.put("type", type.toString());
+        data.put("latitude", location.getLatitude());
+        data.put("longitude", location.getLongitude());
+        data.put("caption", caption);
+
+        return functions.getHttpsCallable("dropPin").call(data).continueWith(task -> {
+            String pid = (String) task.getResult().getData();
+            Pin newPin = new Pin(caption, auth.getUid(), type, content,
+                    new GeoPoint(location.getLatitude(), location.getLongitude()));
+            pins.put(pid, newPin);
```

Comment:
Just confirming, the idea here is to keep a local cached version (i.e. `HashMap<Pid, Pin>`) and then append the local set with remote Pins when needed?

Reply:
Yeah, that's the idea. We'd use the cached version where possible to eliminate load times and reduce server requests, and thus costs.